redwood-lazylayout-view/src/main/kotlin/app/cash/redwood/lazylayout/view/ViewLazyList.kt

```diff
@@ -184,71 +200,46 @@ internal open class ViewLazyList private constructor(
         }
       }
 
-    override fun getItemCount(): Int = items.size
+    override fun getItemCount(): Int = processor.size
 
-    override fun getItemViewType(position: Int): Int {
-      return if (items[position] != null) VIEW_TYPE_ITEM else VIEW_TYPE_PLACEHOLDER
-    }
+    override fun getItemViewType(position: Int): Int = VIEW_TYPE_ITEM
 
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
       val container = FrameLayout(parent.context)
       // [onBindViewHolder] is invoked before the default layout params are set, so
       // [View.getLayoutParams] will be null unless explicitly set.
       container.layoutParams = (parent as RecyclerView).layoutManager!!.generateDefaultLayoutParams()
-      return when (viewType) {
-        VIEW_TYPE_PLACEHOLDER -> ViewHolder.Placeholder(container)
-        VIEW_TYPE_ITEM -> ViewHolder.Item(container)
-        else -> error("Unrecognized viewType: $viewType")
-      }
+      return ViewHolder(container)
     }
 
     override fun onBindViewHolder(holder: ViewHolder, position: Int) {
       lastItemHeight = holder.itemView.height
-      val layoutParams = if (crossAxisAlignment == CrossAxisAlignment.Stretch) {
-        FrameLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT)
-      } else {
-        FrameLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT)
-      }
-      layoutParams.apply {
-        gravity = when (crossAxisAlignment) {
-          CrossAxisAlignment.Start -> Gravity.START
-          CrossAxisAlignment.Center -> Gravity.CENTER
-          CrossAxisAlignment.End -> Gravity.END
-          CrossAxisAlignment.Stretch -> Gravity.START
-          else -> throw AssertionError()
-        }
-      }
-      when (holder) {
-        is ViewHolder.Placeholder -> {
-          if (holder.container.childCount == 0) {
-            val placeholder = placeholders.takeOrNull()
-            if (placeholder != null) {
-              placeholder.value.layoutParams = layoutParams
-              holder.container.addView(placeholder.value)
-              holder.itemView.updateLayoutParams { height = WRAP_CONTENT }
-            } else if (holder.container.height == 0) {
-              // This occurs when the ViewHolder has been freshly created, so we set the container
-              // to a non-zero height so that it's visible.
-              holder.itemView.updateLayoutParams { height = lastItemHeight }
```

Comment:
Big regression here . . . I need to add a hook to create placeholders on-demand.

Reply:
Tracking here: https://github.com/cashapp/redwood/issues/1571